shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -55,63 +85,103 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
+    resetState();
+    int parallelism = 1;
+    if (cl.hasOption(PARALLELISM_OPTION.getLongOpt())) {
+      String parellismOption = cl.getOptionValue(PARALLELISM_OPTION.getLongOpt());
+      parallelism = Integer.parseInt(parellismOption);
+    }
     String[] args = cl.getArgs();
     for (String path : args) {
       AlluxioURI inputPath = new AlluxioURI(path);
       runWildCardCmd(inputPath, cl);
     }
-    return 0;
-  }
-
-  /**
-   * Persists a file or directory currently stored only in Alluxio to the UnderFileSystem.
-   *
-   * @param filePath the {@link AlluxioURI} path to persist to the UnderFileSystem
-   */
-  private void persist(AlluxioURI filePath) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(filePath);
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
-        try {
-          persist(newPath);
-        } catch (Exception e) {
-          errorMessages.add(e.getMessage());
-        }
-      }
-      if (errorMessages.size() != 0) {
-        throw new IOException(Joiner.on('\n').join(errorMessages));
-      }
-    } else if (status.isPersisted()) {
-      System.out.println(filePath + " is already persisted");
-    } else {
-      try {
-        FileSystemUtils.persistFile(mFileSystem, filePath);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new RuntimeException(e);
-      } catch (TimeoutException e) {
-        throw new RuntimeException(e);
+    mTotalFilesToPersist = mToPersist.size();
+    System.out.println("Found " + mTotalFilesToPersist + " files to persist.");
+    ExecutorService service = Executors.newFixedThreadPool(parallelism);
+    List<Future<Void>> persistCalls = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      persistCalls.add(service.submit(new PersistCallable()));
+    }
+    try {
+      for (Future<Void> call : persistCalls) {
+        call.get();
       }
-      System.out.println("persisted file " + filePath + " with size " + status.getLength());
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
     }
+    return 0;
   }
 
   @Override
   public String getUsage() {
-    return "persist <path> [<path> ...]";
+    return "persist [--parallelism <#>] <path> [<path> ...]";
   }
 
   @Override
   public String getDescription() {
     return "Persists files or directories currently stored only in Alluxio to the "
         + "UnderFileSystem.";
   }
+
+  /**
+   * Clears the queue to persist and internal variables. This class is not designed to be used
+   * concurrently or from an existing incomplete state.
```

Comment:
I feel like we're adding unnecessary complexity to the usage and lifecycle for this class. Could we instead pass parameters the the methods that need them, or create a new class in the `run()` method which manages the state?

Reply:
I don't think we can avoid adding state to this class if we want to leverage the `runWildCardCmd` and `runPlainPath` pattern?

Reply:
`runWildCardCmd` makes sense when each file is processed independently. For more complicated commands like this, it adds little value while complicating the control flow

Reply:
Good point, I've changed the code to be self contained.